components/Banner/src/MDCBannerView.m

```diff
@@ -0,0 +1,474 @@
+    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor
+        constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
+    self.trailingButtonConstraintTop =
+        [self.trailingButton.topAnchor constraintEqualToAnchor:self.leadingButton.bottomAnchor
+                                                      constant:kButtonVerticalIntervalSpace];
+    self.trailingButtonConstraintTrailing = [self.trailingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+  }
+}
+
+- (void)deactivateAllConstraints {
+  self.imageViewConstraintLeading.active = NO;
+  self.imageViewConstraintTopSmall.active = NO;
+  self.imageViewConstraintTopLarge.active = NO;
+  self.imageViewConstraintBottom.active = NO;
+  self.textLabelConstraintLeadingWithMargin.active = NO;
+  self.textLabelConstraintLeadingWithImage.active = NO;
+  self.textLabelConstraintTrailing.active = NO;
+  self.textLabelConstraintTop.active = NO;
+  self.textLabelConstraintCenterY.active = NO;
+  self.buttonContainerConstraintLeading.active = NO;
+  self.buttonContainerConstraintLeadingWithTextLabel.active = NO;
+  self.buttonContainerConstraintLeadingWithTextLabelGreater.active = NO;
+  self.buttonContainerConstraintTrailing.active = NO;
+  self.buttonContainerConstraintTopWithMargin.active = NO;
+  self.buttonContainerConstraintTopWithImageViewGreater.active = NO;
+  self.buttonContainerConstraintTopWithTextLabel.active = NO;
+  self.buttonContainerConstraintTopWithTextLabelGreater.active = NO;
+  self.buttonContainerConstraintBottom.active = NO;
+  self.leadingButtonConstraintLeading.active = NO;
+  self.leadingButtonConstraintTop.active = NO;
+  self.leadingButtonConstraintTrailing.active = NO;
+  self.leadingButtonConstraintCenterY.active = NO;
+  self.leadingButtonConstraintTrailingWithTrailingButton.active = NO;
+  self.trailingButtonConstraintBottom.active = NO;
+  self.trailingButtonConstraintTop.active = NO;
+  self.trailingButtonConstraintTrailing.active = NO;
+}
+
+- (void)dealloc {
+  [self removeObservers];
+}
+
+#pragma mark - Property Getter
+
+- (BOOL)hasImage {
+  return self.imageView.image != nil && !self.imageView.hidden;
+}
+
+#pragma mark - KVO
+
+- (void)addObserversToSubviews {
+  [self.imageView addObserver:self
+                   forKeyPath:kMDCBannerViewImageViewImageKeyPath
+                      options:NSKeyValueObservingOptionNew
+                      context:nil];
+}
+
+- (void)removeObservers {
+  [self.imageView removeObserver:self forKeyPath:kMDCBannerViewImageViewImageKeyPath];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary *)change
+                       context:(void *)context {
+  if (!context) {
+    if ([keyPath isEqualToString:kMDCBannerViewImageViewImageKeyPath]) {
+      self.imageView.hidden = (change[NSKeyValueChangeNewKey] == [NSNull null]);
+    }
+  }
```

Comment:
Any KVO code needs to be treated as a last resort, so it's important that you've shown you've exhausted all alternative solutions. Do you feel that hiding the image view like this is necessary behavior, or can we get away with empty image views hanging around as non-hidden empty views?

I believe I recall mention that other UIKit components work in a similar way, please list them here so that if we do go this route we can document which UIKit components are justifying this behavior.

---

Assuming we do go this route, please follow the pattern for KVO used in MDCNavigationBar. Notably it is healthy practice to:

1. kick KVO events to the main thread to reduce the likelihood of off-main-thread UIKit modifications
2. Define your context as a `static char *const`.
3. Invoke super when you don't recognize the context.

https://github.com/material-components/material-components-ios/blob/424a02538bd07b293e9c86020143fadf25c223fd/components/NavigationBar/src/MDCNavigationBar.m#L413-L438

Reply:
It is a very good point which I was also having questions on and unsure about the solutions.

I am not a fan of KVO, so I definitely want to minimize the usage of it in this class. 

> Do you feel that hiding the image view like this is necessary behavior, or can we get away with empty image views hanging around as non-hidden empty views?
I believe I recall mention that other UIKit components work in a similar way, please list them here so that if we do go this route we can document which UIKit components are justifying this behavior.

Yes, we can. I just removed it in 8a5cc283a6b098cd5247404766ede46d078d66db. The behavior I wanted to mimic was `setImage:forState:`, which UIKit allows developers to `setImage:nil forState:..` to set image to be empty and re-layout the button. I think one way to do it is provide a new API (`setImageHidden:(BOOL)hidden`) similar to that, so I can directly update status for constraints. I didn't do that in effort of minimizing the number of APIs at the beginning, but you feel it would be a good way. I think we can add that.

And we can also provide API to hide `trailingButton` by `setTrailingButtonHidden:(BOOL)hidden` instead of `// TODO: KVO trailingButton.hidden and call `setNeedsUpdateConstraints:`

Adding these APIs would also facilitate to integrate to state system in the future if needed.

How do you feel about it? @jverkoey

Reply:
I don't fully understand the proposed changes here, so I'll try to echo back what I believe the assumptions we're working with here are:

1. If a client manually hides one of the buttons, we won't know that the visible state has changed and layout will not be updated accordingly, resulting in an incorrect layout.
2. We do not expect clients to know to invalidate the layout of the banner view if they hide one of the buttons.
3. In order to encourage good behavior, it is proposed that we add an API that clients can use instead of directly setting the button to hidden. This new API would also call setNeedsUpdateConstraints.

Does this match your chain of assumptions?

Reply:
> If a client manually hides one of the buttons, we won't know that the visible state has changed and layout will not be updated accordingly, resulting in an incorrect layout.
Yes, unless we add KVO for the `hidden` property and `setNeedUpdateContraints` in the handler of KVO.

> We do not expect clients to know to invalidate the layout of the banner view if they hide one of the buttons.
Right

> In order to encourage good behavior, it is proposed that we add an API that clients can use instead of directly setting the button to hidden. This new API would also call setNeedsUpdateConstraints.
It is exactly what I proposed:)

Reply:
It sounds then like the options ahead of us are to either add a new API or implement KVO, both of which are additive changes in terms of API or behavior.

For now, I'd like to encourage a conservative approach where we do neither, and instead rely on documenting that if someone hides a button then they need to invalidate the layout (assuming this is a requirement).